### ISSUE TITLE
Support for query parameters

### ISSUE THREAD

**Issue by philippefutureboy** [cube-js/cube.js#481]
**Is your feature request related to a problem? Please describe.**
Given some records with a date field, we want to do aging of the records with regards to the current date or another specified anchor date.
Example:

* Given a Patient table, with a birthDate time field, we would like to be able to group Patients by age slices (0-18 years, 19-30 years, etc.), and be able to filter specifically on this age or age group. Example code:

```
ageGroup: {
      sql: `
      CASE
       -- replace NOW() calls with the query param
        WHEN TIMESTAMPDIFF(YEAR, ${Patient}.birth_date, NOW()) <= 18 THEN '0-18'
        WHEN TIMESTAMPDIFF(YEAR, ${Patient}.birth_date, NOW()) <= 30 THEN '19-30'
        WHEN TIMESTAMPDIFF(YEAR, ${Patient}.birth_date, NOW()) <= 45 THEN '31-45'
        WHEN TIMESTAMPDIFF(YEAR, ${Patient}.birth_date, NOW()) <= 60 THEN '46-60'
        ELSE '61+'
      END
      `,
      type: `string`
    },
```

* Given a Charges & Payments table, we want to be able to categorize the uncollected Charges based on the aging of the Charge (0-30 days, 31-60 days, etc.) and be able to filter for a certain category in order to produce a drilled-down version of the aging category with respect to which Patient has outsanding Charges within the said category.

**Describe alternatives you've considered**
A potential solution is to precalculate the various differences between any two dates of interest and store the result in a materialized table, "CalendarAging" table.
In the first case, the CalendarAging table has a number of records equal to the cross join of the past 100 years worth of dates with the past ~ up to 5 years of last dates, for a total of ~67 million records. The sheer number of produced records could be enough to reduce the usability of such a solution both performance and costs wise (depending on the rates & frequency of access).
Pavel pointed out that the standard solution in the industry is to reduce the number of possible use cases available to end-users by creating a much smaller, finite set of cases. Example: "age at the end of current year" or "age at the end of 2015".

**Describe the solution you'd like**
As Pavel pointed out on DM in Slack, query parameters break the principle of invariant data representation leveraged by OLAP cubes, thereby undermining the ability to create pre-aggregations. However, there are multiple use cases where this may be an important feature to have nevertheless, as Pavel will discuss below.
An optimal solution would allow us to reference query parameters in the dimension & measure definitions, as follows:

```
ageGroup: {
      sql: `
      CASE
        WHEN TIMESTAMPDIFF(YEAR, ${Patient}.birth_date, {{params.anchorDate}}) <= 18 THEN '0-18'
        WHEN TIMESTAMPDIFF(YEAR, ${Patient}.birth_date, {{params.anchorDate}}) <= 30 THEN '19-30'
        WHEN TIMESTAMPDIFF(YEAR, ${Patient}.birth_date, {{params.anchorDate}}) <= 45 THEN '31-45'
        WHEN TIMESTAMPDIFF(YEAR, ${Patient}.birth_date, {{params.anchorDate}}) <= 60 THEN '46-60'
        ELSE '61+'
      END
      `,
      type: `string`
    },
```

Interesting questions to consider:

* How can this feature be documented in such a way that the developers are aware of the potential performance costs of using parameters OR are aware of how to leverage parameters in a way that will enable Cube.js to build a key based on params which will be used in pre-aggregation?
* Should the query params format require that the type of the values passed be explicitly defined?
* Should the query params format allow objects and deeply nested properties?
* Should the query params functionality allow more complex use of templated languages such as mustache or should it be reserved to simple use-cases such as SQL native parameters (named, positional)?

There you go!
Thanks again for the good work and may Cube.js keep shining! 🚀

**Comment by paveltiunov** [comment]
@philippefutureboy Hey Philippe! Thanks for posting this one! Yep. We're considering support of query parameters which in fact will produce separate version of cube per each set of parameters.

**Comment by JFCeron** [comment]
Hi. I'm facing a perfect use case for this (measure SQL should depend on a query parameter). Should I expect the feature to become available soon, or just build a workaround?

**Comment by paveltiunov** [comment]
@JFCeron and whoever is stuck with this issue, please post your use cases here. We can discuss workarounds, if any and it'll help properly design this feature.

**Comment by opera33** [comment]
@paveltiunov I have a use case to actually choose table based on time granularity, the source data is using Apache Doris, where we had pre-aggregated table into minute, hourly, daily, would it be possible for Cube to query from different table provided the time dimension is passed in or query parameter, the column are the same across the table, but in different time granularity.

I understand that pre-aggregation can be done on Cube, but the pre-aggregated table had addressed late data, and it can be updated real-time. or if there's any workaround this.

Basically it shall be SELECT * FROM table_<minute>, when minute granularity passed in.

Thanks

**Comment by alexchrnv** [comment]
**Upvoting**

@paveltiunov 
Pavel, first of all, cheers for the top-tier product you and the community are building!

Here's another use case
### **Is your feature request related to a problem? Please describe.**
Given a Sales table with the following columns:

- created_at
- product_name
- price
- qty

we want to perform an [ABC analysis](https://en.wikipedia.org/wiki/ABC_analysis), which implies setting thresholds for A and B classes
SQL definition for such a cube would be long, but trivial, containing filters like:
```
CASE
    WHEN cumulative_percentage <= 0.8 THEN 'A'
    WHEN cumulative_percentage <= 0.95 THEN 'B'
    ELSE 'C'
END as category
```

### **Describe alternatives you've considered**
1. FILTER_PARAMS

```
CASE
    WHEN cumulative_percentage <= ${FILTER_PARAMS.sales.a_threshold.filter("a_threshold")} THEN 'A'
    WHEN cumulative_percentage <= ${FILTER_PARAMS.sales.b_threshold.filter("b_threshold")} THEN 'B'
    ELSE 'C'
END as category
```

The problem is that 'a_threshold' filter is added to the generated wrapping query, breaking the whole query. I do understand this is not intended use of FILTER_PARAMS.

2. Passing arbitrary data in the HTTP headers and inject them into securityContext via extendContext

cube.js contents:
```
module.exports = {
  extendContext: (req) => {
      return {
          securityContext: {
              ...req.securityContext,
              a_threshold: req.headers['x-a-threshold'] ? req.headers['x-a-threshold'] : null,
              b_threshold: req.headers['x-b-threshold'] ? req.headers['x-b-threshold'] : null
          }
      } 
  },
// remaining cube.js code
}
```

and injecting these values into the cube sql definition:
```
CASE
    WHEN cumulative_percentage <= `${SECURITY_CONTEXT.a_threshold.unsafeValue() ? JSON.parse(SECURITY_CONTEXT.a_threshold.unsafeValue()) : 0.8}` THEN 'A'
    WHEN cumulative_percentage <= `${SECURITY_CONTEXT.b_threshold.unsafeValue() ? JSON.parse(SECURITY_CONTEXT.b_threshold.unsafeValue()) : 0.95}` THEN 'B'
    ELSE 'C'
END as category
```

The second workaround leads to the following problems:
- When using JavaScript SDK, we have to re-instanciate CubeApi for each query to be able to provide headers within CubeApiOptions. It's actually a not too bad option in this particular case (ABC analysis with 1 query), but picture a an enterprice-grade BI dashboard with multiple widgets on the page, and a query for each widget is triggered on each little thing (filter update, cross-filtering click, etc). We'd instanciate CubeApi like dozens of times for each user interaction. The problem can solved by:
  - Forking and modifying the JavaScript SDK, so that load() method accepts headers without reinstantiating CubeApi;
  - Implementing our own cube REST client instead of cube JavaScript SDK.

### **Describe the solution you'd like**
As @philippefutureboy described above:
Implement 'params' object in the cube definition to be used like this:
```
cube(`sales`, {
  sql: '
    CASE
          WHEN cumulative_percentage <=${PARAMS.a_threshold.sanitize()} THEN 'A'
          WHEN cumulative_percentage <= ${PARAMS.b_threshold.sanitize()} THEN 'B'
          ELSE 'C'
      END as category
  ',
  dimensions: {
  // dimensions definition
  },
  measures: {
  // measures definition
  },
  params: {
    a_threshold: {
      type: 'number',
      validation: <some validation rules>
    },
    b_threshold: {
      type: 'number',
      validation: <some validation rules>
    }
  }
}
```

By validation rules I assume some object defining numeric range or regex validation.
By .sanitize() I assume basic sanitization to avoid an sql injection.

Thanks for considering this feature, and all the best wishes to Cube.js team! ❤️

**Comment by igorlukanin** [comment]
@alexchrnv Thanks for a very elaborate comment/suggestion. I think you have a great understanding of the possible solutions space. Just a couple comments from me:
* `COMPILE_CONTEXT` should be used instead of `SECURITY_CONTEXT` ([docs](https://cube.dev/docs/reference/data-model/context-variables#compile_context))
* You can use the common mechanism to pass the security context via the JWT token.
* Instead of recreating the client library instance for each request, I would suggest to maintain a pool/dictionary of client library instances (I doubt they change for every query).

**Comment by alexchrnv** [comment]
@igorlukanin Igor, thanks for the suggestion. My thoughts:


> * `COMPILE_CONTEXT` should be used instead of `SECURITY_CONTEXT` ([docs](https://cube.dev/docs/reference/data-model/context-variables#compile_context))

Got it, SECURITY_CONTEXT is deprecated, thanks!



> * You can use the common mechanism to pass the security context via the JWT token.

Technically, I can, but re-issueing the the JWT token for each input update doesn't seem to be a desired solution in this case. 
The case is that the user updates a parameter (a_threshold and b_threshold in this case. Same as a custom input parameter in Tableau or DataLens) on the frontend, and our app should run .load() SDK method so that an SQL query is executed with this parameter.



> * Instead of recreating the client library instance for each request, I would suggest to maintain a pool/dictionary of client library instances (I doubt they change for every query).

Still doesn't seem like a clean solution. We would need to destroy and re-create clients in the pool for each parameter update. Picture another case - mortgage calculator, where interest rate, installment and period are parameters which should be injected into the SQL query (not in outer filters of the generated SQL as cube filters do). The whole UX of such a dashboard is literally tweaking these 3 parameters, which would require to re-instanciate cube-core SDK on each change in order to pass the parameters values in the HTTP headers.

So still really looking forward to this improvement!
